File: package-python/python/perpetual/iv.py
"""Boosted Instrumental Variable (BoostIV) estimator.

Implements a two-stage least-squares approach using gradient boosting to
estimate causal effects in the presence of endogeneity.
"""

from typing import Any, Dict, Iterable, Optional, Union

import numpy as np
from typing_extensions import Self

from perpetual.perpetual import IVBooster as CrateIVBooster
from perpetual.utils import convert_input_array, convert_input_frame


class BraidedBooster:
    """Two-stage instrumental-variable estimator powered by gradient boosting.

    Stage 1 regresses the treatment on the instruments, and Stage 2 regresses
    the outcome on the predicted treatment and covariates.  Both stages are
    fitted using Perpetual's self-generalizing boosting.
    """

    def __init__(
        self,
        treatment_objective: str = "SquaredLoss",
        outcome_objective: str = "SquaredLoss",
        stage1_budget: float = 0.5,
        stage2_budget: float = 0.5,
        num_threads: Optional[int] = None,
        monotone_constraints: Union[Dict[Any, int], None] = None,
        force_children_to_bound_parent: bool = False,
        missing: float = np.nan,
        allow_missing_splits: bool = True,
        create_missing_branch: bool = False,
        terminate_missing_features: Optional[Iterable[Any]] = None,
        missing_node_treatment: str = "None",
        log_iterations: int = 0,
        quantile: Optional[float] = None,
        reset: Optional[bool] = None,
        categorical_features: Union[Iterable[int], Iterable[str], str, None] = "auto",
        timeout: Optional[float] = None,
        iteration_limit: Optional[int] = None,
        memory_limit: Optional[float] = None,
        stopping_rounds: Optional[int] = None,
        max_bin: int = 256,
        max_cat: int = 1000,
        interaction_constraints: Optional[list[list[int]]] = None,
    ):
        """
        Boosted Instrumental Variable (BoostIV) Estimator.

        Implements a 2-Stage Least Squares (2SLS) approach using Gradient Boosting.

        Parameters
        ----------
        treatment_objective : str, default="SquaredLoss"
            Objective for Stage 1 (Treatment Model). e.g., ``"SquaredLoss"`` or ``"LogLoss"``.
        outcome_objective : str, default="SquaredLoss"
            Objective for Stage 2 (Outcome Model). e.g., ``"SquaredLoss"``.
        stage1_budget : float, default=0.5
            Fitting budget for Stage 1. Higher values allow more boosting rounds.
        stage2_budget : float, default=0.5
            Fitting budget for Stage 2. Higher values allow more boosting rounds.
        num_threads : int, optional
            Number of threads to use during training and prediction.
        monotone_constraints : dict, optional
            Constraints mapping feature indices/names to -1, 1, or 0.
        force_children_to_bound_parent : bool, default=False
            Whether to restrict children nodes to be within the parent's range.
        missing : float, default=np.nan
            Value to consider as missing data.
        allow_missing_splits : bool, default=True
            Whether to allow splits that separate missing from non-missing values.
        create_missing_branch : bool, default=False
            Whether to create a separate branch for missing values (ternary trees).
        terminate_missing_features : iterable, optional
            Features for which missing branches are always terminated when
            ``create_missing_branch`` is True.
        missing_node_treatment : str, default="None"
            How to handle weights for missing nodes. Options: ``"None"``,
            ``"AssignToParent"``, ``"AverageLeafWeight"``, ``"AverageNodeWeight"``.
        log_iterations : int, default=0
            Logging frequency (every N iterations). 0 disables logging.
        quantile : float, optional
            Target quantile when using ``"QuantileLoss"``.
        reset : bool, optional
            Whether to reset the model or continue training on subsequent fits.
        categorical_features : iterable or str, default="auto"
            Feature indices or names to treat as categorical.
        timeout : float, optional
            Time limit for fitting in seconds.
        iteration_limit : int, optional
            Maximum number of boosting iterations.
        memory_limit : float, optional
            Memory limit for training in GB.
        stopping_rounds : int, optional
            Number of rounds without improvement before stopping.
        max_bin : int, default=256
            Maximum number of bins for feature discretization.
        max_cat : int, default=1000
            Maximum unique categories before a feature is treated as numerical.
        interaction_constraints : list of list of int, optional
            Groups of feature indices allowed to interact.
        """
        self.treatment_objective = treatment_objective
        self.outcome_objective = outcome_objective
        self.stage1_budget = stage1_budget
        self.stage2_budget = stage2_budget
        self.num_threads = num_threads
        self.monotone_constraints = monotone_constraints
        self.force_children_to_bound_parent = force_children_to_bound_parent
        self.missing = missing
        self.allow_missing_splits = allow_missing_splits
        self.create_missing_branch = create_missing_branch
        self.terminate_missing_features = terminate_missing_features
        self.missing_node_treatment = missing_node_treatment
        self.log_iterations = log_iterations
        self.quantile = quantile
        self.reset = reset
        self.categorical_features = categorical_features
        self.timeout = timeout
        self.iteration_limit = iteration_limit
        self.memory_limit = memory_limit
        self.stopping_rounds = stopping_rounds
        self.max_bin = max_bin
        self.max_cat = max_cat
        self.interaction_constraints = interaction_constraints

        # Placeholder for standardized constraints and categorical features.
        # These are usually calculated in fit(), but we need to pass something to __init__
        # for CrateIVBooster if we follow the PerpetualBooster pattern of recreating it.
        # However, IVBooster's fit currently uses the same instance.

        self.booster = CrateIVBooster(
            treatment_objective=treatment_objective,
            outcome_objective=outcome_objective,
            stage1_budget=stage1_budget,
            stage2_budget=stage2_budget,
            max_bin=max_bin,
            num_threads=num_threads,
            monotone_constraints={},  # Passed during fit in PerpetualBooster
            interaction_constraints=interaction_constraints,
            force_children_to_bound_parent=force_children_to_bound_parent,
            missing=missing,
            allow_missing_splits=allow_missing_splits,
            create_missing_branch=create_missing_branch,
            terminate_missing_features=set(),
            missing_node_treatment=missing_node_treatment,
            log_iterations=log_iterations,
            quantile=quantile,
            reset=reset,
            categorical_features=set(),
            timeout=timeout,
            iteration_limit=iteration_limit,
            memory_limit=memory_limit,
            stopping_rounds=stopping_rounds,
        )

    def fit(self, X, Z, y, w) -> Self:
        """
        Fit the IV model.

        Parameters
        ----------
        X : array-like
            Covariates (Controls).
        Z : array-like
            Instruments.
        y : array-like
            Outcome variable.
        w : array-like
            Treatment received.
        """
        # Feature conversion
        features_x, flat_x, rows_x, cols_x, _, _ = convert_input_frame(X, "auto", 1000)
        self.feature_names_in_ = features_x

        # Z conversion (features)
        _, flat_z, rows_z, cols_z, _, _ = convert_input_frame(Z, "auto", 1000)

        if rows_x != rows_z:
            raise ValueError("X and Z must have the same number of rows.")

        # Treatment conversion
        w_, _ = convert_input_array(w, self.treatment_objective)

        # Outcome conversion
        y_, _ = convert_input_array(y, self.outcome_objective)

        self.booster.fit(flat_x, rows_x, cols_x, flat_z, rows_z, cols_z, y_, w_)
        return self

    def predict(self, X, w_counterfactual) -> np.ndarray:
        """
        Predict Outcome given X and a counterfactual W.

        Parameters
        ----------
        X : array-like
            Covariates.
        w_counterfactual : array-like
            Treatment value to simulate.

        Returns
        -------
        preds : ndarray
            Predicted Outcome.
        """
        _, flat_x, rows_x, cols_x, _, _ = convert_input_frame(X, "auto", 1000)

        w_cf_, _ = convert_input_array(
            w_counterfactual, self.treatment_objective
        )  # Use same encoding logic as W

        preds = self.booster.predict(flat_x, rows_x, cols_x, w_cf_)
        return preds

    def to_json(self) -> str:
        """Serialize model to JSON string."""
        return self.booster.json_dump()

    @classmethod
    def from_json(cls, json_str: str) -> "BraidedBooster":
        """Deserialize model from JSON string."""
        obj = cls.__new__(cls)
        obj.booster = CrateIVBooster.from_json(json_str)
        params = obj.booster.get_params()
        obj.treatment_objective = params["treatment_objective"]
        obj.outcome_objective = params["outcome_objective"]
        obj.stage1_budget = params["stage1_budget"]
        obj.stage2_budget = params["stage2_budget"]
        return obj

File: src/causal/iv.rs
//! Instrumental Variable (BoostIV) Estimator
//!
//! Implements a two-stage least-squares (2SLS) approach using gradient boosting
//! to estimate causal effects in the presence of endogeneity.
use crate::booster::config::{CalibrationMethod, MissingNodeTreatment};
use crate::booster::core::PerpetualBooster;
use crate::constraints::ConstraintMap;
use crate::data::Matrix;
use crate::errors::PerpetualError;
use crate::objective_functions::objective::Objective;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

/// Boosted Instrumental Variable (BoostIV) Estimator.
///
/// Implements a 2-Stage Least Squares (2SLS) approach using Gradient Boosting.
///
/// # Model Structure
///
/// * **Stage 1 (Treatment Model)**: Estimates the treatment intensity based on instruments and covariates.
///   $$ \hat{W} = f_1(X, Z) + \epsilon_1 $$
///   - Objective: `SquaredLoss` (Regression) or `LogLoss` (Classification/Propensity).
///
/// * **Stage 2 (Outcome Model)**: Estimates the outcome based on the *predicted* treatment and covariates.
///   $$ Y = f_2(X, \hat{W}) + \epsilon_2 $$
///   - The causal effect is derived from the dependence of $Y$ on $\hat{W}$.
///
#[derive(Serialize, Deserialize)]
pub struct IVBooster {
    /// Stage 1 model that predicts the treatment from instruments + covariates.
    pub treatment_model: PerpetualBooster,
    /// Stage 2 model that predicts the outcome from predicted treatment + covariates.
    pub outcome_model: PerpetualBooster,
    /// Budget allocated to the stage 1 (treatment) model.
    pub stage1_budget: f32,
    /// Budget allocated to the stage 2 (outcome) model.
    pub stage2_budget: f32,
}

impl IVBooster {
    /// Create a new `IVBooster` instance for Instrumental Variable estimation.
    ///
    /// # Arguments
    ///
    /// * `treatment_objective` - Objective for the stage 1 (treatment) model (e.g., `LogLoss` or `SquaredLoss`).
    /// * `outcome_objective` - Objective for the stage 2 (outcome) model (e.g., `SquaredLoss`).
    /// * `stage1_budget` - Learning budget for the treatment model.
    /// * `stage2_budget` - Learning budget for the outcome model.
    /// * `max_bin` - Maximum number of bins for feature discretization.
    /// * `num_threads` - Optional number of threads.
    /// * `monotone_constraints` - Optional monotonicity constraints.
    /// * `interaction_constraints` - Optional interaction constraints.
    /// * `force_children_to_bound_parent` - Whether to bound child predictions by parent's range.
    /// * `missing` - Missing value representation.
    /// * `allow_missing_splits` - Whether to allow splits isolating missing values.
    /// * `create_missing_branch` - Whether to create explicit missing branches (ternary trees).
    /// * `terminate_missing_features` - Features where missing branches should terminate early.
    /// * `missing_node_treatment` - Strategy for handling missing value splits.
    /// * `log_iterations` - Logging frequency.
    /// * `seed` - Random seed.
    /// * `quantile` - Target quantile for quantile regression.
    /// * `reset` - Whether to reset or continue training on fit.
    /// * `categorical_features` - Features to treat as categorical.
    /// * `timeout` - Hard limit for fitting time in seconds.
    /// * `iteration_limit` - Hard limit for number of iterations.
    /// * `memory_limit` - Memory limit in GB.
    /// * `stopping_rounds` - Number of rounds for early stopping.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        treatment_objective: Objective,
        outcome_objective: Objective,
        stage1_budget: f32,
        stage2_budget: f32,
        max_bin: u16,
        num_threads: Option<usize>,
        monotone_constraints: Option<ConstraintMap>,
        interaction_constraints: Option<Vec<Vec<usize>>>,
        force_children_to_bound_parent: bool,
        missing: f64,
        allow_missing_splits: bool,
        create_missing_branch: bool,
        terminate_missing_features: HashSet<usize>,
        missing_node_treatment: MissingNodeTreatment,
        log_iterations: usize,
        seed: u64,
        quantile: Option<f64>,
        reset: Option<bool>,
        categorical_features: Option<std::collections::HashSet<usize>>,
        timeout: Option<f32>,
        iteration_limit: Option<usize>,
        memory_limit: Option<f32>,
        stopping_rounds: Option<usize>,
    ) -> Result<Self, PerpetualError> {
        // Initialize Stage 1 Booster
        let treatment_model = PerpetualBooster::new(
            treatment_objective,
            stage1_budget,
            f64::NAN, // Auto-calc base score
            max_bin,
            num_threads,
            monotone_constraints.clone(),
            interaction_constraints.clone(),
            force_children_to_bound_parent,
            missing,
            allow_missing_splits,
            create_missing_branch,
            terminate_missing_features.clone(),
            missing_node_treatment,
            log_iterations,
            seed,
            quantile,
            reset,
            categorical_features.clone(),
            timeout,
            iteration_limit,
            memory_limit,
            stopping_rounds,
            CalibrationMethod::default(),
        )?;

        // Initialize Stage 2 Booster
        let outcome_model = PerpetualBooster::new(
            outcome_objective,
            stage2_budget,
            f64::NAN,
            max_bin,
            num_threads,
            monotone_constraints,
            interaction_constraints,
            force_children_to_bound_parent,
            missing,
            allow_missing_splits,
            create_missing_branch,
            terminate_missing_features,
            missing_node_treatment,
            log_iterations,
            seed,
            quantile,
            reset,
            categorical_features,
            timeout,
            iteration_limit,
            memory_limit,
            stopping_rounds,
            CalibrationMethod::default(),
        )?;

        Ok(IVBooster {
            treatment_model,
            outcome_model,
            stage1_budget,
            stage2_budget,
        })
    }

    /// Fit the IV Model.
    ///
    /// # Arguments
    /// * `X` - Covariates (Controls).
    /// * `Z` - Instruments.
    /// * `y` - Partial Outcome.
    /// * `w` - Treatment received.
    pub fn fit(&mut self, x: &Matrix<f64>, z: &Matrix<f64>, y: &[f64], w: &[f64]) -> Result<(), PerpetualError> {
        // --- Stage 1: Treatment Model ---
        // Predict W using X and Z.
        // We need to concatenate X and Z features.
        // For simplicity/performance in this initial version, we assume the user
        // passes a combined matrix for Stage 1 or we construct it.
        // Let's assume for now we construct a lightweight combined view or new matrix.
        // TODO: Optimize this concatenation (maybe use ColumnarMatrix).

        // Construct Stage 1 Features: [X, Z]
        // This is expensive if we copy.
        // For now, let's assume the user handles feature engineering and passes `stage1_features` directly?
        // Or strictly strictly implement 2SLS logic here.

        // Let's implement the concatenation logic for correctness first, optimization later.
        let rows = x.rows;
        let x_cols = x.cols;
        let z_cols = z.cols;
        let total_cols = x_cols + z_cols;
        // Efficient Column-Major concatenation
        // Assumes input matrices are standard contiguous column-major (which they are by Matrix definition)
        let mut stage1_data = Vec::with_capacity(x.data.len() + z.data.len());
        stage1_data.extend_from_slice(x.data);
        stage1_data.extend_from_slice(z.data);

        let matrix_stage1 = Matrix::new(&stage1_data, rows, total_cols);

        // Fit Stage 1
        self.treatment_model.fit(&matrix_stage1, w, None, None)?;

        // Predict W_hat
        let w_hat = self.treatment_model.predict(&matrix_stage1, true);

        // --- Stage 2: Outcome Model ---
        // Predict Y using X and W_hat.
        // Stage 2 Features: [X, W_hat]
        let mut stage2_data = Vec::with_capacity(x.data.len() + w_hat.len());
        stage2_data.extend_from_slice(x.data);
        stage2_data.extend_from_slice(&w_hat);

        let matrix_stage2 = Matrix::new(&stage2_data, rows, x_cols + 1);

        // Fit Stage 2
        self.outcome_model.fit(&matrix_stage2, y, None, None)?;

        Ok(())
    }

    /// Predict Outcome given X (and potentially new Z, though usually we want E[Y|do(W)]).
    /// For IV, "prediction" is tricky. Usually we want the structural function f(X, W).
    ///
    /// * `X` - Covariates
    /// * `w_counterfactual` - Treatment value to simulate (e.g., 1.0 or 0.0)
    pub fn predict(&self, x: &Matrix<f64>, w_counterfactual: &[f64]) -> Vec<f64> {
        let rows = x.rows;
        let x_cols = x.cols;

        // Construct Stage 2 Features with counterfactual W
        // Stage 2 Features with counterfactual: [X, W_cf]
        let mut stage2_data = Vec::with_capacity(x.data.len() + rows);
        stage2_data.extend_from_slice(x.data);

        // Handle W_counterfactual (broadcast if scalar, else extend)
        if w_counterfactual.len() == 1 {
            let val = w_counterfactual[0];
            stage2_data.resize(stage2_data.len() + rows, val);
        } else {
            if w_counterfactual.len() != rows {
                // If this happens, we might panic or should handle error, but signature doesn't allow error.
                // Assuming caller provides correct length as per doc logic implies match.
                // Fallback to safe iteration/extend if needed, or just extend.
                stage2_data.extend_from_slice(w_counterfactual);
            } else {
                stage2_data.extend_from_slice(w_counterfactual);
            }
        }

        let matrix_stage2 = Matrix::new(&stage2_data, rows, x_cols + 1);

        // Predict using Outcome Model
        self.outcome_model.predict(&matrix_stage2, true)
    }
}
